spec/formatting.md

```diff
@@ -1,24 +1,113 @@
-# WIP DRAFT MessageFormat 2.0 Formatting Behaviour
+# DRAFT MessageFormat 2.0 Formatting Behaviour
 
 ## Introduction
 
 This document defines the behaviour of a MessageFormat 2.0 implementation
 when formatting a message for display in a user interface, or for some later processing.
 
-The document is part of the MessageFormat 2.0 specification,
-the successor to ICU MessageFormat, henceforth called ICU MessageFormat 1.0.
+Message formatting is defined in terms of the following stages:
 
-## Literal Resolution
+- **_Resolution_** determines the value of a part of the message,
+  such as an _expression_, with reference to the current _formatting context_.
+  This may include multiple steps,
+  such as looking up the value of a variable and calling formatting functions.
+  The resolved value is not necessarily in the shape it will finally take,
+  but is "formattable", i.e. it contains everything required by the eventual formatting.
 
-The resolved value of _text_, _literal_ and _nmtoken_ tokens
+  > For example, the resolved value of the _expression_ `{|0.40| :number style=percent}`
+  > could be an object such as
+  >
+  > ```
+  > { value: Number('0.40'),
+  >   formatter: NumberFormat(locale, { style: 'percent' }) }
+  > ```
+  >
+  > The shape of this value is completely implementation-dependent,
+  > and different implementations may choose to perform different levels of resolution.
+
+- **_Pattern Selection_** determines which of a message's _patterns_ is formatted.
+  For a message with no _selectors_, this is simple as there is only one _pattern_.
+  With _selectors_, this will depend on their _resolution_.
+
+- **_Formatting_** takes the resolved values of the selected _pattern_,
+  and formats them in the desired shape.
+  This specification only defines formatting messages as a single concatenated string,
+  but implementations SHOULD provide formatters for additional shapes
+  as appropriate for their setting.
```

Comment:
This is what I was mentioning in the call. We should start to think in terms of `formatToParts`. We can't say how the implementation works or what the name of objects/classes/data structures are. But we can describe formatting as resolving the _message_ into a logical sequence of (um, erm) values. 

---

A formatted message as a whole has some properties associated with it: locale, base direction, and a sequence of "parts"

Each "part" also has a set of properties. The parts are in a logically ordered sequence or array. Each part has a locale and a base direction property. Additional properties MAY be defined by the implementation.

There are two kinds of "part": a "literal part" and an "expression" part.

Each "literal part" consists of a string.

An "expression part" can be resolved to a sequence of zero or more "literal parts".

The string output of a message is the concatenated sequence of resolved literal parts. 

> Here is a simple terrible example:
> Inputs:
> | What | Value | Description |
> |---|---|---|
> | Locale | ar-AE | Locale to use for formatting |
> | `date` | `2023-06-19` | Data value passed to formatter |
>
> Message:
> 
> `{The example date is {$date :datetime skeleton=yMMMd}}`
>
> Output:
> ```
> {
>    "locale": "ar-AE",
>    "direction": "ltr",
>    "parts": [
>        {
>            "type": "literal",
>            "locale": "ar-AE",
>            "direction": "ltr",
>            "value": "The example date is "
>        },
>        {
>            "type": "expression",
>            "locale": "ar-AE",
>            "direction": "rtl",
>            "value": [
>                   { "type": "literal", "locale":"ar-AE", "dir": "rtl", "name": "day", "19" },
>                   { "type": "literal", "locale":"ar-AE", "dir": "rtl", "name": "separator", " " },
>                   { "type": "literal", "locale":"ar-AE", "dir": "rtl", "name": "month", "يونيو" },
>                   { "type": "literal", "locale":"ar-AE", "dir": "rtl", "name": "separator", " " },
>                   { "type": "literal", "locale":"ar-AE", "dir": "rtl", "name": "year", "2023" },
>            ]
>        },
>    ]
> }
> ```

Callers can consume the sequence of parts in order to perform additional processing, such as markup. The above example might be formatted into an HTML context thusly:

> ```html
> <p lang="ar-AE" dir="ltr">The example date is 
>     <span dir="rtl" id="date">19 <em>يونيو</em> 2023</span></p>
> ```

Reply:
> Each "literal part" consists of a string.
> 
> An "expression part" can be resolved to a sequence of zero or more "literal parts".

This is not always true. For one example, consider this message:
```
{This is my image: {$img}. Isn't it pretty?}
```

When formatting this, the value of `$img` could be a representation of the image itself, rather than any sequence of literal parts.

Next, consider this message:
```
{This is my image: {flower.png :image}. Isn't it pretty?}
```

Here, the message doesn't include any variables, but it does make use of a custom `:image` function, which could format as a representation of the image that is similarly non-stringifiable.

In other words, we cannot make any assumptions about the shape of external variables or the return values of custom functions.

Reply:
Message format can have parts that are objects.

Literal parts of a message are *always* strings by definition.

Expressions can, as you note, represent objects and these might not be immediately stringable. But the existence of string resolution (note your text does this!!) means that all expressions can ultimately be represented as a sequence of literals. It is tempting to want to make it a list of literals *or* expressions. But ultimately what my text says is that you **can** resolve the deepest nesting of an expression to a string.

The string might be `<img src="flower.png">` or `img/png:base64gooHere` or something. And users do not have to force the object to become a string (they can peek at the expression "type" and get the image, for example).

I'm open to a lot of change here: my proposal above is basically the back of my cocktail napkin while thinking about message resolution. Allowing "shapes" other than string is fine, but having our standard require that one be able to produce a character sequence means that everything can ultimately call `toString`.

To @macchiati's point, "shape" is kind of a vague word. Perhaps:

```suggestion
- **_Formatting_** takes the resolved values of the selected _pattern_,
  and returns the formatted result for the _message_.
  This specification defines formatting of each _message_ as a _string_.
 Implementations MAY return a _message_ using a different, locally appropriate, 
 data type (such as an attributed string) or as a logical sequence of 
 values as appropriate for that implementation.
 
 > For example, an implementation might choose to return an interstitial
 > object so that the caller can "decorate" portions of the formatted value.
 > The `NumberFormatter` class in ICU4J, for example, returns a `FormattedNumber`
 > object, so a _pattern_ such as `{This is my number {42 :number}}` might return
 > the character sequence `This is my number ` followed by a `FormattedNumber`
 > object representing the value `42` in the current locale.
```

Reply:
I've taken some of the suggested language and incorporated it into the Formatting section, rather than this Introduction. Also moved out & refactored the resolution example slightly.